### ISSUE TITLE
[Bug]: GPT-OSS-120B gpt-oss MXFP4 on SM121 (Blackwell DGX Spark): Marlin kernel generates wrong first Harmony token, producing null content/reasoning

### ISSUE THREAD

**Issue by ohsono** [vllm-project/vllm#37030]
### Environment

- **vLLM Version**: 0.16.0rc2.dev559+g4bf19f9bf
- **CUDA**: 13.0
- **GPU**: NVIDIA B200 / DGX Spark (SM121)
- **Model**: openai/gpt-oss-120b

**Server command:**
```bash
vllm serve openai/gpt-oss-120b \
  --quantization mxfp4 \
  --dtype bfloat16 \
  --kv-cache-dtype fp8 \
  --max-model-len 2048 \
  --gpu-memory-utilization 0.70 \
  --max-num-seqs 2 \
  --enable-chunked-prefill \
  --enable-prefix-caching \
  --load-format fastsafetensors
```

---

### 🐛 Bug Description

When serving `openai/gpt-oss-120b` with MXFP4 quantization on a **single SM121 GPU** (NVIDIA B200 / DGX Spark), all chat completions return `content: null` and `reasoning: null` despite `completion_tokens` showing that tokens were generated.

**Request:**
```python
client.chat.completions.create(
    model="gpt-oss-120b",
    messages=[{"role": "user", "content": "What is 2+2?"}],
    max_tokens=512,
)
```

**Response:**
```json
{
  "choices": [{
    "message": {
      "role": "assistant",
      "content": null,
      "reasoning": null,
      "tool_calls": []
    },
    "finish_reason": "length"
  }],
  "usage": {
    "prompt_tokens": 74,
    "completion_tokens": 512,
    "total_tokens": 586
  }
}
```

512 tokens were generated but none surfaced as content or reasoning.

---

### Root Cause

**The Harmony prompt rendered for `gpt-oss-120b` ends with:**
```
<|start|>assistant        (tokens: 200006, 173781)
```

The model must generate `<|channel|>` (token **200005**) as its first completion token to begin a valid Harmony channel header (`<|channel|>analysis<|message|>...`).

**Instead, on SM121 with MXFP4 + Marlin fallback, the model generates token `181886` = `'farbe'` (German: "color") — a completely wrong non-Harmony text token.**

Verified via `return_token_ids: true`:
```json
"prompt_token_ids": [..., 200006, 173781],
"token_ids":        [181886, 87, 314, ...]
```

Decoded: `'farbex = 0.5\nWe need to produce the next token...'`

The Harmony `StreamableParser` immediately raises `HarmonyError` on token 181886 because it expects `<|start|>` (200006) or `<|channel|>` (200005), not a regular text token. The parser returns an empty result → `reasoning: null`, `content: null`.

**Server log evidence (before HarmonyError fix):**
```
HarmonyError during output parsing, returning partial result:
  Unexpected token 63248 while expecting start token 200006
```

---

### Why This Happens

SM121 (Blackwell) does **not** have native hardware FP4 support. The vLLM server logs confirm:

```
WARNING: Your GPU does not have native support for FP4 computation but FP4
quantization is being used. Weight-only FP4 compression will be used
leveraging the Marlin kernel. This may degrade performance for compute-heavy
workloads.
```

The `csrc/quantization/gptq_marlin/` directory contains only SM80-targeted kernels:
```
sm80_kernel_bfloat16_fe2m1f_bfloat16.cu
sm80_kernel_bfloat16_fe4m3fn_bfloat16.cu
...
```

No SM121-specific Marlin kernels exist. The SM80 Marlin MXFP4 kernel running on SM121 produces incorrect logits for the first assistant token, causing the model to sample a garbage text token instead of the Harmony control token `<|channel|>`.

---

### Relationship to Existing Issues

- **#31422** ([Bug]: SM120/SM100: gpt-oss gibberish with tp=2) — related but different: that issue is TP=2 + prefix caching. This bug reproduces on **single GPU, tp=1**, with prefix caching enabled or disabled.

---

### Secondary Bug: System Message Causes Hallucinated Reasoning

A second bug causes the model to produce hallucinated/off-topic reasoning when a **system message** is included in the request. With the broken prompt, the model sometimes generates tokens in the `analysis` channel, but the content is completely wrong (e.g., discussing Codeforces problems when asked "What is 2+2?", or outputting Russian/Unicode characters).

**Root cause:** `_make_request_with_harmony` was passing system-role messages directly to `parse_chat_inputs_to_harmony_messages`, which does not handle the `system` role in Harmony encoding format. The raw system tokens corrupt the Harmony token sequence, misdirecting the model's reasoning.

This is fixed in the linked PR.

---

### Steps to Reproduce

1. SM121 GPU (NVIDIA B200/DGX Spark)
2. Start vLLM with `openai/gpt-oss-120b --quantization mxfp4`
3. Send any chat completion request
4. Observe: `content: null`, `reasoning: null`, `finish_reason: length`
5. Enable `return_token_ids: true` — first completion token is not `200005` (`<|channel|>`)

---

### Expected Behavior

First completion token should be `200005` (`<|channel|>`), followed by valid Harmony channel content, producing non-null `reasoning` and/or `content`.

---

### Proposed Fix

Add SM121-targeted Marlin MXFP4 kernels in `csrc/quantization/gptq_marlin/` or route SM121 through FlashInfer's native MXFP4 path (if supported) rather than the SM80 Marlin fallback.

---

### Workaround

Remove `--quantization mxfp4` and use `--dtype bfloat16` (unquantized). This avoids the Marlin kernel entirely. Note: 120B model in bfloat16 requires ~240 GB VRAM.

**Comment by mohit8632273-ops** [comment]
Assign this issue to me

**Comment by ohsono** [comment]
Tracking related work for visibility:

- **PR #41028** ([Kernel] OAITritonExperts MXFP4: include SM 12.x in supported device range) addresses the device gate `_triton_kernel_moe_supports_current_device()` that caps at SM < 11.0, excluding SM 12.0/12.1. This is likely a contributing factor to the null content/reasoning on DGX Spark.
- **PR #31607** fixes the V1 engine startup crashes and CUTLASS op fallbacks on SM 12.1, which surfaces this issue more reliably by letting the engine actually run on the hardware.
- **PR #47350** handles `HarmonyError` in the streaming `process_chunk` path, which catches the symptom (null content with no diagnostic) gracefully.

The root cause — Marlin kernel generating the wrong first Harmony token on MXFP4/SM121 — still requires a kernel-level fix. Marking this to avoid auto-close.

**Comment by ohsono** [comment]
## Live validation on SM 12.1 (GB10 / DGX Spark) — `openai/gpt-oss-120b`

Ran a targeted test suite against a local `gpt-oss-120b` instance on SM 12.1 hardware covering the scenarios from this issue and its related PRs (#47350, #41028, #31607). All 8/8 pass.

**Environment:** SM 12.1 (GB10, 121 GiB VRAM), vLLM v0.23.1rc1, CUDA 13.0, torch 2.11.0+cu130

| # | Test | Related | finish | content null? | Result |
|---|------|---------|--------|--------------|--------|
| 1 | Baseline — no system msg (`What is 2+2?`) | #37030 repro | `stop` | ✅ no | PASS |
| 2 | System msg `str` content — hallucination regression | #37030 / Bug 1 | `stop` | ✅ no | PASS |
| 3 | System msg `list[{type,text}]` format | #37030 / Bug 1 | `stop` | ✅ no | PASS |
| 4 | Streaming — simple Q, no HarmonyError crash | #47350 | `stop` | ✅ no | PASS |
| 5 | Streaming + system msg | #47350 + #37030 | `stop` | ✅ no | PASS |
| 6 | Multi-turn with system msg | #37030 | `stop` | ✅ no | PASS |
| 7 | Streaming multi-turn | #47350 + #37030 | `stop` | ✅ no | PASS |
| 8 | `reasoning_effort=low` | #37030 | `stop` | ✅ no | PASS |

**Key observations:**
- The original `content: null, reasoning: null` symptom from the issue (512 tokens generated but nothing surfaced) is **not reproducible** on current main. The null-output bug is fixed — system message extraction (`extract_instructions_from_messages`, merged via #34268/#43590) and `HarmonyError` handling in `flush()`/`parse()`/`parse_delta()` (partial, #47350 covers the remaining `process_chunk` gap) are in place.
- All streaming paths complete with `finish_reason: stop` and non-null content — no `HarmonyError` escapes.
- `reasoning_effort=low` routes correctly and returns non-null content.

**Still open:** The deeper Marlin kernel first-token issue (wrong token generated under MXFP4 quantization, causing the Harmony parser to receive an unexpected token at position 0) is a kernel-level fix. PR #41028 addresses the MXFP4 device gate for SM 12.x which is a prerequisite. The `process_chunk` gap is tracked in PR #47350.
